src/duration/mod.rs

```diff
@@ -0,0 +1,356 @@
+//! This module contains a common [`Duration`] struct which is able to parse
+//! human-readable duration formats, like `2y 2h 20m 42s` or`15d 2m 2s`. It
+//! additionally implements many required traits, like [`Derivative`],
+//! [`JsonSchema`], [`Deserialize`][serde::Deserialize], and
+//! [`Serialize`][serde::Serialize].
+//!
+//! Furthermore, it implements [`Deref`], which enables us to use all associated
+//! functions of [`std::time::Duration`] without re-implementing the public
+//! functions on our own type.
+//!
+//! All operators should opt for [`Duration`] instead of the plain
+//! [`std::time::Duration`] when dealing with durations of any form, like
+//! timeouts or retries.
+
+use std::{
+    fmt::{Display, Write},
+    num::ParseIntError,
+    ops::{Deref, DerefMut},
+    str::FromStr,
+};
+
+use derivative::Derivative;
+use schemars::JsonSchema;
+use strum::IntoEnumIterator;
+use thiserror::Error;
+
+mod serde_impl;
+pub use serde_impl::*;
+
+#[derive(Debug, Error, PartialEq)]
+pub enum DurationParseError {
+    #[error("invalid input, either empty or contains non-ascii characters")]
+    InvalidInput,
+
+    #[error("failed to parse duration fragment")]
+    FragmentError(#[from] DurationFragmentParseError),
+}
+
+#[derive(Clone, Copy, Debug, Derivative, Hash, PartialEq, PartialOrd, JsonSchema)]
+pub struct Duration(std::time::Duration);
+
+impl FromStr for Duration {
+    type Err = DurationParseError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let input = s.trim();
+
+        // An empty or non-ascii input is invalid
+        if input.is_empty() || !input.is_ascii() {
+            return Err(DurationParseError::InvalidInput);
+        }
+
+        // Let's split up individual parts separated by a space
+        let parts: Vec<&str> = input.split(' ').collect();
+
+        // Parse each part as a DurationFragment and extract the final duration
+        // of each fragment in milliseconds
+        let values: Vec<u128> = parts
+            .iter()
+            .map(|p| p.parse::<DurationFragment>())
+            .map(|r| r.map(|f| f.millis()))
+            .collect::<Result<Vec<_>, DurationFragmentParseError>>()?;
```

Comment:
I think I'd rather parse each parse each fragment as a `std::time::Duration` and then sum them up (I also made this mistake in handling seconds too directly in my parser).

Reply:
I can add a `From<DurationFragment> for std::time:.Duration` impl.

Reply:
Changed in 9e9257d